enhancements/microshift/microshift-supportability-tools.md

```diff
@@ -0,0 +1,365 @@
+the APIs are not the same, resources are not the same, and the output format
+is required to follow `must-gather` format.
+
+Since none of the available plugins fulfills MicroShift's needs, we will need a
+new one.
+
+Requirements for the `sos` plugin follow:
+* Not require usage of other tools than what is already present in the host.
+  No additional images or componentes shall be pulled when performing a report.
+* Have `must-gather` format to use already existing debug tools.
+* Retrieve MicroShift specific resources only, not covered by other plugins.
+  These include cluster resources, MicroShift configuration, MicroShift
+  version and MicroShift logs. Some specifics, like ovnk will require an
+  additional plugin, as openshift does. Other components, such as CRI-O, have
+  their own plugins.
+* Provide a single command pre-configured for full report gathering.
+  In order to have a single and simple command to gather the report, provide a
+  profile configuration file for `sos` to run all relevant plugins in one shot.
+* Must not leak sensitive data: IPs, hostnames, passwords, certificates, etc.
+  The /var/lib/microshift directory holds certificates and etcd data, therefore
+  it must be skipped.
+
+Existing debugging tools in use by CEE include [omg](https://github.com/kxr/o-must-gather)
+and [omc](https://github.com/gmeghnag/omc). These provide a CLI like experience
+in terms of getting information from a live cluster, ressembling `oc` commands,
+but coming from a must-gather report. Since they were born out of must-gather,
+they expect the same file structure layout. The must-gather file system
+hierarchy is driven by `oc adm inspect` commands, gathering cluster resources
+data that these tools read upon later.
+
+To achieve the same results the most simple approach is to include the `oc`
+binary with MicroShift as a dependency. The plugin can then call
+`oc adm inspect` to produce the output required. There is no API between the
+plugin and MicroShift, easing maintainability and allowing enough flexibility
+for the future.
+
+### Workflow Description
+> _Disclaimer:_ Workflows described here illustrate manual processes.
+
+Assuming from now on MicroShift has already been installed in a R4E/RHEL based
+system and there has been some kind of issue.
+
+1. MicroShift administrator notices an issue/wants to get a system report.
+1. The administrator will log in to the failing host.
+1. The administrator will perform the debug report creation procedure:
+```bash
+sudo sos report --config-file /etc/microshift/report.conf
+```
+
+For CI jobs the same procedure applies. This shall be automated upon failure.
+
+#### Variation [optional]
+N/A
+
+### API Extensions
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+N/A
+
+### Risks and Mitigations
+There are four main risks:
+1. Upgrades and incompatible changes.
+1. `sos` release cadence.
+1. CEE having a different tool.
+1. Exposing sensitive data.
+
+#### Upgrades
```

Comment:
I'm missing descriptions how the tool will work in specific scenarios:
1. in a working microshift
2. if microshift fails
3. if microshift upgrade fails
4. if base os fails
5. if base of upgrade fails
6. if ceritifactes fail
7. else? 

What are the assumptions about the data we can gather when microshift is running or not, which data are coming from base os and which require a running microshift?

Reply:
> I'm missing descriptions how the tool will work in specific scenarios:
> 
> 1. in a working microshift
> 2. if microshift fails
> 3. if microshift upgrade fails
> 4. if base os fails
> 5. if base of upgrade fails
> 6. if ceritifactes fail
> 7. else?
> 
> What are the assumptions about the data we can gather when microshift is running or not, which data are coming from base os and which require a running microshift?

Those are great questions and we could adjust the sos plugin to collect the specific data based on that scenarios. I want to just add that this is the advantage of the sosreport, where sosreport will collect data even when the MicroShift is not running. The key in that scenario are the service logs, which will be collected by the sosreport even if `oc adm inspect` will fail. From support perspective, this is the key information needed to move forward with resolving the issues.
If base os or upgrade fails, than it is RHEL troubleshooting and not MicroShitf. 

@pamoedom, I know that we don't want to collect the `/var/lib/microshift` in sosreport, but it might be beneficial to collect some data from that directory, like certificates and kubeconfigs (without private information). Sosreport's openshift plugin already does that perfectly. What do you think?

Reply:
> Those are great questions and we could adjust the sos plugin to collect the specific data based on that scenarios. I want to just add that this is the advantage of the sosreport, where sosreport will collect data even when the MicroShift is not running. The key in that scenario are the service logs, which will be collected by the sosreport even if `oc adm inspect` will fail. From support perspective, this is the key information needed to move forward with resolving the issues. If base os or upgrade fails, than it is RHEL troubleshooting and not MicroShitf.

Even if MicroShift itself is not running, containers might be (as those rely on CRI-O) and their logs can still be collected by this plugin as well.


> @pamoedom, I know that we don't want to collect the `/var/lib/microshift` in sosreport, but it might be beneficial to collect some data from that directory, like certificates and kubeconfigs (without private information). Sosreport's openshift plugin already does that perfectly. What do you think?

The openshift plugin is scrubbing all secret data after collection: https://github.com/sosreport/sos/blob/main/sos/report/plugins/openshift.py#L405-L435
Is it useful to have kubeconfigs and certificates if all sensitive data (which is the certs) is scrubbed? You would be left with a bunch of yaml that does not really tell you anything.